selene_sdk/samplers/random_positions_sampler.py

```diff
@@ -282,6 +321,118 @@ def _retrieve(self, chrom, position):
+            self.samples[mode]["proportions"] = \
+                [chr_len / self.samples[mode]["length"] for chr_len in whole_proportions[mask]]
+
+        return
+
+    # Update `samples_per_chrom` to map a chromosomes to the number of times it
+    # should be sampled in a `sample_size` set, given that each chrom's
+    # ratio is given by the `proportions` list. `
+    def _samples_from_proportion(self, samples_per_chrom, mode):
+        for i in range(self.samples[mode]["size"]):
+            chr = np.random.choice(self.samples[mode]["chroms"],
+                                   p=self.samples[mode]["proportions"])
+            samples_per_chrom[chr] += 1
+
+    # Make a dictionary of chromosomes which maps to how many elements from chrom
+    # we sample. Determined probabilistically and weighted by proportion.
+    def _samples_per_chrom(self):
+        self._set_proportions_and_chroms()
+        samples_per_chrom = defaultdict(int)
+        for mode in ("train", "validation", "test"):
+            self._samples_from_proportion(samples_per_chrom, mode)
+
+        return samples_per_chrom
+
+    # Setup `self.chroms`, `self.chroms_starts`, `self.chroms_ends`, `self.genome_length`
+    def _init_chroms(self):
+        tot_len = 0
+        for chrom, len_chrom in self.reference_sequence.get_chr_lens():
+            self.chroms_info["Total"] = np.append(self.chroms_info["Total"], chrom)
+            self.chroms_info["Starts"] = np.append(self.chroms_info["Starts"], tot_len)
+            self.chroms_info["Ends"] = np.append(self.chroms_info["Ends"], tot_len + len_chrom)
+            tot_len += len_chrom
+            if chrom in self.test_holdout:
+                self.samples["test"]["length"] += len_chrom
+            elif chrom in self.validation_holdout:
+                self.samples["validation"]["length"] += len_chrom
+            else:
+                self.samples["train"]["length"] += len_chrom
+
+        self.genome_length = tot_len
+
+    # Generate a sample for a given chromosome with `samples_per_chrom` elements.
+    def _generate_chrom_sample(self, samples_per_chrom, len_chrom, tot_len):
+        chrom_range = np.arange(tot_len, tot_len + len_chrom)
+        curr_sample = np.random.choice(chrom_range,
+                                    size=(1, samples_per_chrom),
+                                    replace=False)[0]
+        return curr_sample
+
+    # Updates train_sample to hold the indices we will sample from
+    # of a hypothetical flattened array of each chromosome laid side by side.
+    # These indices can then be converted into (chrom, position) pairs to sample.
+    def _generate_sample(self):
+        self._init_chroms()
+
+        tot_len = 0
+        train_counter = 0
+        validation_counter = 0
+        test_counter = 0
+
+        samples_per_chrom = self._samples_per_chrom()
+        for chrom, len_chrom in self.reference_sequence.get_chr_lens():
+            _samples_per_chrom = samples_per_chrom[chrom]
```

Comment:
super confusing variable names... shouldn't have a private `_samples_per_chrom` and a `samples_per_chrom` right above